pyro/infer/predictive.py

```diff
@@ -255,36 +268,25 @@ def log_like_helper(*args, **kwargs) -> torch.Tensor:
             for name, site in trace.nodes.items()
             if site["type"] == "sample" and site["is_observed"]
         }
-        # now trace it and extract the likelihood from observed sites
-        predictive = Predictive(
-            model, posterior_samples, guide, num_samples, (), parallel)
         # use vectorized trace
         if parallel:
             log_like = dict()
+            # now trace it and extract the likelihood from observed sites
+            predictive = Predictive(
+                model, posterior_samples, guide, num_samples, (), parallel)
             trace = predictive.get_vectorized_trace(*args, **kwargs)
             for obs_name, obs_val in observed.items():
                 obs_site = trace.nodes[obs_name]
                 log_like[obs_name] = obs_site["fn"].log_prob(obs_val).detach().cpu()
         # iterate over samples from posterior if model can't be vectorized
         else:
-            all_samples = {
-                k: v for k, v in predictive(*args, **kwargs).items()
-                if k not in observed}
-            log_like = {obs_name: [] for obs_name in observed}
-            for i in range(num_samples):
-                param_sample = {k: v[i] for k, v in all_samples.items()}
-                cond_model = pyro.condition(model, data=param_sample)
-                traced_model = poutine.trace(cond_model)
-                trace = traced_model.get_trace(*args, **kwargs)
-                for obs_name, obs_val in observed.items():
-                    obs_site = trace.nodes[obs_name]
-                    ll_sample = obs_site["fn"].log_prob(obs_val).detach().cpu()
-                    log_like[obs_name].append(ll_sample)
-            # each sample is a tensor in a list; stack all the elements into
-            # a single tensor for each site
-            log_like = {
-                obs_name: torch.stack(ll_list, dim=0)
-                for obs_name, ll_list in log_like.items()
+            site_shapes = {
+                name: (num_samples,) + trace.nodes[name]["value"].shape
+                for name in observed
             }
+            log_like = _predictive_sequential(
```

Comment:
In case `guide` is not None (like in `Predictive` implementation), I think you need
```
posterior_samples = _predictive(guide, {}, num_samples, return_sites=None,
                                parallel=False, model_args=args, model_kwargs=kwargs)
```

It seems that you need to add some logic to get `num_samples` or assertations as in `Predictive` constructor. Do you think that supporting `guide`, `num_samples` makes things complicated?